core/store/src/flat/storage.rs

```diff
@@ -679,4 +780,393 @@ mod tests {
+            shard_uid,
+            FlatStorageStatus::Ready(FlatStorageReadyStatus { flat_head: chain.get_block(0) }),
+        );
+        store_helper::set_flat_state_value(
+            &mut store_update,
+            shard_uid,
+            vec![1],
+            Some(FlatStateValue::value_ref(&[0])),
+        );
+        store_update.commit().unwrap();
+
+        for i in 1..num_blocks as BlockHeight {
+            let mut store_update = store.store_update();
+            let changes = if i % 3 == 0 {
+                // Add a change.
+                FlatStateChanges::from([(vec![1], Some(FlatStateValue::value_ref(&[i as u8])))])
+            } else {
+                // No changes.
+                FlatStateChanges::from([])
+            };
+
+            // Simulates `Chain::save_flat_state_changes()`.
+            let prev_block_with_changes = if changes.0.is_empty() {
+                store_helper::get_prev_block_with_changes(
+                    &store,
+                    shard_uid,
+                    chain.get_block(i).hash,
+                    chain.get_block(i).prev_hash,
+                )
+                .unwrap()
+            } else {
+                None
+            };
+            let delta = FlatStateDelta {
+                changes,
+                metadata: FlatStateDeltaMetadata {
+                    block: chain.get_block(i),
+                    prev_block_with_changes,
+                },
+            };
+            tracing::info!(?i, ?delta);
+            store_helper::set_delta(&mut store_update, shard_uid, &delta);
+            store_update.commit().unwrap();
+        }
+
+        let flat_storage_manager = FlatStorageManager::new(store.clone());
+        flat_storage_manager.create_flat_storage_for_shard(shard_uid).unwrap();
+        let flat_storage = flat_storage_manager.get_flat_storage_for_shard(shard_uid).unwrap();
+
+        // 2. Check that the chunk_view at block i reads the value of key &[1] as &[round_down_to_a_multiple_of_3(i)]
+        for i in 0..num_blocks as BlockHeight {
+            let block_hash = chain.get_block_hash(i);
+            let blocks = flat_storage.get_blocks_to_head(&block_hash).unwrap();
+            let chunk_view = flat_storage_manager.chunk_view(shard_uid, block_hash).unwrap();
+            let value = chunk_view.get_value(&[1]).unwrap();
+            tracing::info!(?i, ?block_hash, ?value, blocks_to_head = ?blocks);
+            assert_eq!(value, Some(FlatStateValue::value_ref(&[((i / 3) * 3) as u8])));
+
+            for block_hash in blocks {
+                let delta = store_helper::get_delta_changes(&store.clone(), shard_uid, block_hash)
+                    .unwrap()
+                    .unwrap();
+                assert!(
+                    !delta.0.is_empty(),
+                    "i: {i}, block_hash: {block_hash:?}, delta: {delta:?}"
+                );
+            }
+        }
+
+        // 3. Simulate moving head according to finalization of blocks.
```

Comment:
I'm not sure if we need to know about finality in scope of storage.
Can we just say instead that we check blocks on distance 2 and explain why there are <= 4 hops? (because I don't fully get it)